src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpRequestTPFatTPFServerWithTranslation.java

```diff
@@ -0,0 +1,215 @@
+		final TPFRequest newReq = new TPFRequestImpl(tp);
+		final TPFResponse res = FederationAccessUtils.performRequest(fedAccessMgr, newReq, fm);
+		final Set<Triple> tripleTranslation = new HashSet<>();
+		for(final Triple i : res.getPayload()) {
+			tripleTranslation.addAll(translateResultTriple(i, newReq.getQueryPattern()));
+		}
+		return tripleTranslation;
+	}
+	
+	protected Set<Triple> handleUnionPattern(final SPARQLUnionPattern up, final FederationAccessManager fedAccessMgr) throws FederationAccessException{
+		final Set<Triple> unionTranslation = new HashSet<>();
+		for(final SPARQLGraphPattern i : up.getSubPatterns()) {
+			if (i instanceof TriplePattern) {
+				unionTranslation.addAll(handleTriplePattern(((TriplePattern) i), fedAccessMgr));
+			} else if (i instanceof SPARQLUnionPattern) {
+				unionTranslation.addAll(handleUnionPattern(((SPARQLUnionPattern) i), fedAccessMgr));
+			} else if (i instanceof SPARQLGroupPattern) {
+				unionTranslation.addAll(handleGroupPattern(((SPARQLGroupPattern) i), fedAccessMgr));
+			} else if (i instanceof BGP) {
+				unionTranslation.addAll(handleBGP(((BGP) i), fedAccessMgr));
+			} else {
+				throw new FederationAccessException(i.toString(), req, fm);
+			}
+		}
+		return unionTranslation;
+	}
+	
+	protected Set<Triple> handleGroupPattern(final SPARQLGroupPattern gp, final FederationAccessManager fedAccessMgr) throws FederationAccessException{
+		List<Set<Triple>> partialTranslations = new ArrayList<>();
+		for(final SPARQLGraphPattern i : gp.getSubPatterns()) {
+			if (i instanceof TriplePattern) {
+				partialTranslations.add(handleTriplePattern((TriplePattern) i, fedAccessMgr));
+			} else if (i instanceof SPARQLUnionPattern) {
+				partialTranslations.add(handleUnionPattern(((SPARQLUnionPattern) i), fedAccessMgr));
+			} else if (i instanceof SPARQLGroupPattern) {
+				partialTranslations.add(handleGroupPattern(((SPARQLGroupPattern) i), fedAccessMgr));
+			} else if (i instanceof BGP) {
+				partialTranslations.add(handleBGP(((BGP) i), fedAccessMgr));
+			} else {
+				throw new FederationAccessException(i.toString(), req, fm);
+			}
+		}
+		Set<Triple> groupTranslation = new HashSet<>();
+		for(final Triple j : partialTranslations.get(0)) {
+			boolean join = true;
+			ListIterator<Set<Triple>> k = partialTranslations.listIterator(1);
+			while(k.hasNext()) {
+				if (!k.next().contains(j)){
+					join = false;
+					break;
+				}
+			}
+			if (join) {
+				groupTranslation.add(j);
+			}
+		}
+		return groupTranslation;
+	}
+	
+	protected Set<Triple> handleBGP(final BGP bgp, final FederationAccessManager fedAccessMgr) throws FederationAccessException{
+		List<Set<Triple>> partialTranslations = new ArrayList<>();
+		for(final TriplePattern i : bgp.getTriplePatterns()) {
+			partialTranslations.add(handleTriplePattern(i, fedAccessMgr));
+		}
+		Set<Triple> bgpTranslation = new HashSet<>();
+		for(final Triple j : partialTranslations.get(0)) {
```

Comment:
I have been thinking a bit about this and I have come up with the following idea, that uses Set functionality and can make this code a bit easier to follow:
First, initialize a set with the resulting triples of evaluating the first triple pattern (e.g. `final Set<Triple> res`).
For each other triple pattern in the bgp, use the function `res.retainAll(otherSet)` that only keeps elements in the intersection.
At the end you'll have the intersection of all the sets of triples.